### ISSUE TITLE
Optionally delete up to possible end in selected region

### ISSUE THREAD

**Issue by zphr** [Fuco1/smartparens#867]
Hello,

first off thank you for your hard work on this great extension! By now I could not work without it!
The following is a feature request to delete the content of the region as far as possible and therefore change the current behavior of rejecting the deletion to a more DWIM style.

The selected region is wrapped with / in the following simple example.

## Start Input
```
if (bla === true) {
  /const ok = well(3000);
  coolFuntction(ok);
}
/|
const blub = 3;
```

## Expected behavior
```
if (bla === true) {
|}

const blub = 3;
```

## Actual behavior

SmartParens rightly tells me that the deletion of the region would produce an unbalanced result.

I am sure I have missed a lot of implications and if this is outright impossible please disregard this feature request. 

Thanks again for this great piece of work!

**Comment by Fuco1** [comment]
Actually, there is an extension of smartparens for supporting evil and they have this exact feature implemented, so it is definitely possible (in vim when you use the `V` line selection mode and then hit `d` it is somehow smart about what it deletes and up to where)

Here's an animation: https://github.com/expez/evil-smartparens#is-it-any-good

I will try to look at their code and steal some of it back :D

/cc @expez Maybe it would be possible to move that functionality into sp core?

**Comment by expez** [comment]
> Maybe it would be possible to move that functionality into sp core?

Evil smartparens is pretty naive.  It just calls `sp-region-ok-p` and then shrinks the region until the predicate says it's OK.  

This was the 'simplest possible thing that could world', but computers are so fast now I've never had to tweak it for performance reasons.

tl;dr, that would be cool, but there isn't much to move :)

**Comment by Fuco1** [comment]
@expez Fair enough :D  I think `sp--region-ok-p` could be extended to return some data in case of error, because as it walks the forms it should have the "last good one" and from that I suppose it would be trivial to calculate the end of the good region.

This sounds intriguing enough that I might actually look into it :D

**Comment by Fuco1** [comment]
@expez I've added `sp-get-region-info` which gives both the status (ok/not ok) and the last good sexp, of which `(sp-get last-good-sexp :end-suf)` should give the extend of the most permissible region to delete.

@zphr There is an option `sp-auto-narrow-kill-region` if you set it to `t` it should do what you want.  It's on a branch `feature/dwim-kill-region` I need to add some tests but it seems it's working fine.

**Comment by zphr** [comment]
@Fuco1 Yeah, works great! Exactly what I wanted. Thank you 👍!
